### PR TITLE
feat(dashboards): Added ignore time picker as options field to the dashboard variables

### PR DESCRIPTION
**Pull request by Avinash-Sanpala**
JIRA: https://new-relic.atlassian.net/browse/NR-170831

### CODE REVIEW

File: pkg/dashboards/dashboards_api.go [hunk 1]
```diff
@@ -2056,15 +2193,6 @@ func (a *Dashboards) DashboardUpdateWithContext(
...
 
-	// If we got errors back, wrap them all up
-	if len(resp.DashboardUpdateResult.Errors) > 0 {
-		errs := fmt.Errorf("query error")
-		for _, err := range resp.DashboardUpdateResult.Errors {
-			errs = fmt.Errorf("%w; %s", errs, err.Description)
-		}
-		return nil, errs
-	}
-
```
Comment (pranav-new-relic) [inline pkg/dashboards/dashboards_api.go:2067]: Can you please revert this change? It looks like this portion of the code was manually added (and not via Tutone) in order to ensure effective error handling, which is why code generation via Tutone would suggest removing these changes while we might actually need them.

https://github.com/newrelic/newrelic-client-go/commit/8f3b6ddebf46210250b6d4db36ce72763920d07d

File: pkg/dashboards/dashboards_api.go [hunk 2]
```diff
@@ -2000,15 +2146,6 @@ func (a *Dashboards) DashboardDeleteWithContext(
...
 
-	// If we got errors back, wrap them all up
-	if len(resp.DashboardDeleteResult.Errors) > 0 {
-		errs := fmt.Errorf("query error")
-		for _, err := range resp.DashboardDeleteResult.Errors {
-			errs = fmt.Errorf("%w; %s", errs, err.Description)
-		}
-		return nil, errs
-	}
-
```
Comment (pranav-new-relic) [inline pkg/dashboards/dashboards_api.go:2011]: Can you please revert this change? It looks like this portion of the code was manually added (and not via Tutone) in order to ensure effective error handling, which is why code generation via Tutone would suggest removing these changes while we might actually need them.

https://github.com/newrelic/newrelic-client-go/commit/8f3b6ddebf46210250b6d4db36ce72763920d07d

File: pkg/dashboards/types.go [hunk 3]
```diff
@@ -258,13 +287,8 @@ type DashboardPageInput struct {
...
 	// The name of the page.
 	Name string `json:"name"`
-
-	// NOTE: The JSON description of the following attribute, "Widgets" has been modified manually
-	// (removal of "omitempty") to facilitate creating pages with no widgets (empty pages).
-	// Please DO NOT regenerate/modify this attribute and its datatype via Tutone (which would add "omitempty" back).
-
 	// A nested block of all widgets belonging to the page.
-	Widgets []DashboardWidgetInput `json:"widgets"`
+	Widgets []DashboardWidgetInput `json:"widgets,omitempty"`
```
Comment (pranav-new-relic) [inline pkg/dashboards/types.go:291]: Can you please revert this change (considering the note in the comments above this line)

This change was made in an older PR in order to remove `omitempty` against widgets, as this field would get omitted from the request passed to the API if no widgets are included in the page that is added in the API request, thereby blocking users from creating dashboards with empty pages. This had to be added manually and not via Tutone, since Tutone currently doesn't support removing `omitempty` against optional fields.

File: pkg/dashboards/types.go [hunk 4]
```diff
@@ -317,13 +341,8 @@ type DashboardUpdatePageInput struct {
...
 	// Page name.
 	Name string `json:"name"`
-
-	// NOTE: The JSON description of the following attribute, "Widgets" has been modified manually
-	// (removal of "omitempty") to facilitate creating pages with no widgets (empty pages).
-	// Please DO NOT regenerate/modify this attribute and its datatype via Tutone (which would add "omitempty" back).
-
 	// Page widgets.
-	Widgets []DashboardWidgetInput `json:"widgets"`
+	Widgets []DashboardWidgetInput `json:"widgets,omitempty"`
```
Comment (pranav-new-relic) [inline pkg/dashboards/types.go:345]: Can you please revert this change (considering the note in the comments above this line)

This change was made in an older PR in order to remove `omitempty` against widgets, as this field would get omitted from the request passed to the API if no widgets are included in the page that is added in the API request, thereby blocking users from creating dashboards with empty pages. This had to be added manually and not via Tutone, since Tutone currently doesn't support removing `omitempty` against optional fields.

File: pkg/dashboards/types.go [hunk 5]
```diff
@@ -439,15 +460,17 @@ type DashboardVariableInput struct {
...
 	Items []DashboardVariableEnumItemInput `json:"items,omitempty"`
 	// Configuration for variables of type NRQL.
-	NRQLQuery *DashboardVariableNRQLQueryInput `json:"nrqlQuery,omitempty"`
+	NRQLQuery DashboardVariableNRQLQueryInput `json:"nrqlQuery,omitempty"`
 	// Variable identifier.
 	Name string `json:"name"`
+	// Options applied to the variable
+	Options DashboardVariableOptionsInput `json:"options,omitempty"`
 	// Indicates the strategy to apply when replacing a variable in a NRQL query.
 	ReplacementStrategy DashboardVariableReplacementStrategy `json:"replacementStrategy,omitempty"`
```
Comment (pranav-new-relic) [inline pkg/dashboards/types.go:475]:
```go
// DashboardVariableInput - Definition of a variable that is local to this dashboard. Variables are placeholders for dynamic values in widget NRQLs.
type DashboardVariableInput struct {
	// [DEPRECATED] Default value for this variable. The actual value to be used will depend on the type.
	DefaultValue *DashboardVariableDefaultValueInput `json:"defaultValue,omitempty"`
	// Default values for this variable. The actual value to be used will depend on the type.
	DefaultValues *[]DashboardVariableDefaultItemInput `json:"defaultValues,omitempty"`
	// Indicates whether this variable supports multiple selection or not. Only applies to variables of type NRQL or ENUM.
	IsMultiSelection bool `json:"isMultiSelection,omitempty"`
	// List of possible values for variables of type ENUM
	Items []DashboardVariableEnumItemInput `json:"items,omitempty"`
	// Configuration for variables of type NRQL.
	NRQLQuery *DashboardVariableNRQLQueryInput `json:"nrqlQuery,omitempty"`
	// Variable identifier.
	Name string `json:"name"`
	// Options applied to the variable
	Options DashboardVariableOptionsInput `json:"options,omitempty"`
	// Indicates the strategy to apply when replacing a variable in a NRQL query.
	ReplacementStrategy DashboardVariableReplacementStrategy `json:"replacementStrategy,omitempty"`
	// Human-friendly display string for this variable.
	Title string `json:"title,omitempty"`
	// Specifies the data type of the variable and where its possible values may come from.
	Type DashboardVariableType `json:"type"`
}
```

File: pkg/dashboards/types.go [hunk 6]
```diff
@@ -403,7 +424,7 @@ type DashboardVariableDefaultItem struct {
 // DashboardVariableDefaultItemInput - Represents a possible default value item.
 type DashboardVariableDefaultItemInput struct {
 	// The value of this default item.
-	Value DashboardVariableDefaultValueInput `json:"value,omitempty"`
+	Value *DashboardVariableDefaultValueInput `json:"value,omitempty"`
 }
```
Comment (pranav-new-relic) [inline pkg/dashboards/types.go:428]:
```golang
// DashboardVariableDefaultItemInput - Represents a possible default value item.
type DashboardVariableDefaultItemInput struct {
	// The value of this default item.
	Value DashboardVariableDefaultValueInput `json:"value,omitempty"`
}
```

File: pkg/dashboards/types.go [hunk 7]
```diff
@@ -403,7 +424,7 @@ type DashboardVariableDefaultItem struct {
 // DashboardVariableDefaultItemInput - Represents a possible default value item.
 type DashboardVariableDefaultItemInput struct {
 	// The value of this default item.
-	Value DashboardVariableDefaultValueInput `json:"value,omitempty"`
+	Value *DashboardVariableDefaultValueInput `json:"value,omitempty"`
```
Comment (pranav-new-relic) [inline pkg/dashboards/types.go:427]:
```suggestion
	Value DashboardVariableDefaultValueInput `json:"value,omitempty"`
```

File: pkg/entities/types.go [hunk 8]
```diff
@@ -7988,6 +7988,8 @@ type DashboardVariable struct {
 	NRQLQuery *DashboardVariableNRQLQuery `json:"nrqlQuery,omitempty"`
 	// Variable identifier.
 	Name string `json:"name,omitempty"`
+	// Options applied to the variable
+	Options DashboardVariableOptionsInput `json:"options,omitempty"`
```
Comment (pranav-new-relic) [inline pkg/entities/types.go:7992]:
```suggestion
	Options DashboardVariableOptions `json:"options,omitempty"`
```

File: pkg/entities/types.go [hunk 9]
```diff
@@ -7996,6 +7998,12 @@ type DashboardVariable struct {
 	Type DashboardVariableType `json:"type,omitempty"`
 }
 
+// DashboardVariableOptionsInput - Options applied to the variable
+type DashboardVariableOptionsInput struct {
+	// Only applies to variables of type NRQL. With this turned on, the time range for the NRQL query will override the time picker on dashboards and other pages. Turn this off to use the time picker as normal.
+	IgnoreTimeRange bool `json:"ignoreTimeRange,omitempty"`
+}
+
```
Comment (pranav-new-relic) [inline pkg/entities/types.go:8006]:
```suggestion
// DashboardVariableOptions - Options applied to the variable
type DashboardVariableOptions struct {
	// Only applies to variables of type NRQL. With this turned on, the time range for the NRQL query will override the time picker on dashboards and other pages. Turn this off to use the time picker as normal.
	IgnoreTimeRange bool `json:"ignoreTimeRange,omitempty"`
}
```

File: pkg/dashboards/dashboard.go [hunk 10]
```diff
@@ -99,6 +99,9 @@ const getDashboardEntityQuery = `query ($guid: EntityGuid!) {
             value
           }
           name
+		  options {
+				ignoreTimeRange
+			}
```
Comment (pranav-new-relic) [inline pkg/dashboards/dashboard.go:104]:
```suggestion
          options {
            ignoreTimeRange
          }
```

just a tinyformatting fix

File: pkg/dashboards/dashboards_api.go (hunk 1 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/dashboards_api.go:2067]: Sure, I will re-check all these automated code blocks

File: pkg/dashboards/dashboards_api.go (hunk 2 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/dashboards_api.go:2011]: same here

File: pkg/dashboards/dashboards_api.go [hunk 11]
```diff
@@ -36,14 +35,6 @@ func (a *Dashboards) DashboardCreateWithContext(
...
 	}
 
-	if len(resp.DashboardCreateResult.Errors) > 0 {
-		errs := fmt.Errorf("query error")
-		for _, err := range resp.DashboardCreateResult.Errors {
-			errs = fmt.Errorf("%w; %s", errs, err.Description)
-		}
-		return nil, errs
-	}
-
```
[the remark this answers is not in the document]
Comment (Avinash-Sanpala) [reply pkg/dashboards/dashboards_api.go:46]: same here

File: pkg/dashboards/types.go (hunk 3 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/types.go:291]: same here

File: pkg/dashboards/types.go (hunk 4 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/types.go:345]: same here

File: pkg/dashboards/types.go (hunk 5 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/types.go:475]: Upon seeing all the above changes I realised there is a gap between tutone generation and manual modification. I will check if these manual changes need to be accommodated

File: pkg/dashboards/types.go (hunk 7 above)
Comment (Avinash-Sanpala) [reply pkg/dashboards/types.go:427]: Upon seeing all the above changes I realised there is a gap between tutone generation and manual modification. I will check if these manual changes need to be accommodated

File: pkg/entities/types.go (hunk 8 above)
Comment (Avinash-Sanpala) [reply pkg/entities/types.go:7992]: Upon seeing all the above changes I realised there is a gap between tutone generation and manual modification. I will check if these manual changes need to be accommodated

File: pkg/entities/types.go (hunk 9 above)
Comment (Avinash-Sanpala) [reply pkg/entities/types.go:8006]: Upon seeing all the above changes I realised there is a gap between tutone generation and manual modification. I will check if these manual changes need to be accommodated

**Comment by pranav-new-relic** [comment]
(8 March, 2024)

image

https://github.com/newrelic/newrelic-client-go/actions/runs/8199765206/job/22425371395?pr=1106

**Review by Nandu-pns — APPROVED** [review]
